pantos/validatornode/restclient.py

```diff
@@ -0,0 +1,269 @@
+    class TransferSignaturePostRequest:
+        """Request data for posting the signature for a cross-chain
+        token transfer to the primary validator node.
+
+        Attributes
+        ----------
+        source_blockchain : Blockchain
+            The transfer's source blockchain.
+        source_transaction_id : str
+            The transfer's transaction ID/hash on the source blockchain.
+        signature : str
+            The transfer signature of the secondary validator node.
+
+        """
+        source_blockchain: Blockchain
+        source_transaction_id: str
+        signature: str
+
+    def post_transfer_signature(self,
+                                request: TransferSignaturePostRequest) -> None:
+        """Post the signature for a cross-chain token transfer to the
+        primary validator node.
+
+        Parameters
+        ----------
+        request : TransferSignaturePostRequest
+            The request data.
+
+        Raises
+        ------
+        PrimaryNodeDuplicateSignatureError
+            If the primary validator node reports the secondary
+            validator node's transfer signature as already existing.
+        PrimaryNodeInvalidSignatureError
+            If the primary validator node reports the secondary
+            validator node's transfer signature as being invalid.
+        PrimaryNodeInvalidSignerError
+            If the primary validator node reports the signer of the
+            transfer signature as not being a validator node on the
+            transfer's destination blockchain.
+        PrimaryNodeUnknownTransferError
+            If the primary validator node does not know any transfer for
+            the given source blockchain and transaction ID/hash.
+        PrimaryNodeClientError
+            If another error occurs during posting the transfer
+            signature.
+
+        """
+        url = (f'{self.__primary_node_url}{_TRANSFER_SIGNATURE_RESOURCE}')
+        extra_info = vars(request) | {'url': url, 'timeout': self.__timeout}
+        json_request = {
+            'source_blockchain_id': request.source_blockchain.value,
+            'source_transaction_id': request.source_transaction_id,
+            'signature': request.signature
+        }
+        response = self.__send_post_request(url, json_request, extra_info)
+        if response.status_code != requests.codes.no_content:
+            extra_info |= {'status_code': response.status_code}
+            json_response = self.__decode_json_response(response, extra_info)
+            response_message = self.__extract_response_message(
+                json_response, extra_info)
+            if (response.status_code == requests.codes.bad_request
+                    and response_message is not None
+                    and 'Invalid signature.' in response_message):
+                raise PrimaryNodeInvalidSignatureError(**extra_info)
+            if response.status_code == requests.codes.forbidden:
```

Comment:
I feel that the 403 Forbidden is not the right status code for submitting signature which already exists.

403 states that the client has no access rights. In this case our client has access rights, but the data submitted already exists on the server.
I feel 409 Conflict might be better here for the Duplicate signature and 403 for Invalid signer, since this is really another entity which tries to submit a signature

Reply:
I agree, status code 409 is more appropriate for a duplicate signature. I've changed it.